### ISSUE TITLE
"block-closing-brace-space-after" should ignore whitespace after newline

### ISSUE THREAD

**Issue by Nirazul** [stylelint/stylelint#1007]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bug:
There seems to be a subtle difference between the allowing and the forbidding state of the rule:
Allowed > a single space is needed
Disallowed > no white space characters, not even newline are allowed.
I'd have expected a white space to be disallowed and a newline to be ok.

> Which rule, if any, is this issue related to?

`block-closing-brace-space-after`

> What CSS is needed to reproduce this issue?

```scss
.a {}
.b {}

/* also: */
.c {}

.d {}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "block-closing-brace-space-after": "never-single-line",
  }
}
```

> Which version of stylelint are you using?

`HEAD`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

gulp-stylelint

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

no

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:
```shell
test.css
  1:6  error  Unexpected whitespace after "}" of a single-line block   block-closing-brace-space-after
```

**Comment by ntwb** [comment]
Confirmed, added failing tests in #1010

**Comment by davidtheclark** [comment]
I do not think this is a bug as described. (Though I'm not sure I fully understand the intent.)

Using a `never` option with a `-space-` rule *is* the way to enforce *no whitespace at all*. We did this because (a) we need a way to enforce no whitespace at all, and (b) we think it aligns with what most people expect in most situations, and (c) it reduces the number of permutations we may have to deal with.

If you do not want a warning in this CSS ...

```css
.a {}
.b {}
```

... I don't think you should use `"block-closing-brace-space-after": "never-single-line"`.

**Comment by Nirazul** [comment]
Maybe you're right. The thing is, that I've tried out different configurations for my files. And what I would stick with, would be `"block-closing-brace-space-after": "always-single-line"` if it would allow space OR newline. But it complains about a missing space, as there is none before the `\n` finishes the line.

**Comment by davidtheclark** [comment]
I'm hoping we can find a config that fits your intentions. So do you want to be able to do both of the following, regardless of what's inside the rules (that is, whether the declaration block is one line or multiple lines)?

```css
a { color: pink; } b { top: 0; }
```

```css
a { color: pink; } 
b { top: 0; }
```

How about this? Would you accept this?

```css
a { 
  color: pink; 
} b { top: 0; }
```

I guess I'm trying to figure out what exactly you're trying to allow and what you're trying to block.

**Comment by Nirazul** [comment]
Basically these are my desired states: 
```scss
.a {
    display: block;
    color: red;
}

.b { color: pink; }
.c { color: green; }
```

It think what I was unsure about, was the usage with SCSS language constructs like if/else:

```scss
@if (true) {
    // do something
} @else {
    // do something else
}
```

So clearly I want a space between `}` and `@else`

**Comment by ntwb** [comment]
CSS:
```css
a {
    display: block;
    color: red;
}

b { color: pink; }
c { color: green; }

// There is a single space following the closing brace on the next two lines
d { color: pink; } 
e { color: green; } 

f { color: pink; } g { color: red; }
```

Rules:
```js
{
  "rules": {
    "block-opening-brace-newline-after": "always-multi-line",
    "block-closing-brace-newline-after": "always-single-line"
  }
}
```


Results in:
```
Line 10, Col 19: Expected single space after "}" of a single-line block (block-closing-brace-newline-after)
Line 11, Col 20: Expected single space after "}" of a single-line block (block-closing-brace-newline-after)
Line 13, Col 19: Expected single space after "}" of a single-line block (block-closing-brace-newline-after)
```

Selectors `a`, `b`, and `c` work as expected, the `d`, `e`, and `f`/`g` do not, rather the warning for each of these is stating `Expected single space after "}"` rather than `Expected new line after "}"`.

**Comment by jeddy3** [comment]
@Nirazul The following is also what `stylelint-config-standard` [enforces](https://github.com/stylelint/stylelint-config-standard/blob/master/index.js#L7-L13):

```css
.a {
    display: block;
    color: red;
}

.b { color: pink; }
.c { color: green; }
```

So, the following config will enforce this:

```
{
  "block-closing-brace-newline-after": "always",
  "block-closing-brace-newline-before": "always-multi-line",
  "block-closing-brace-space-before": "always-single-line",
  "block-opening-brace-newline-after": "always-multi-line",
  "block-opening-brace-space-after": "always-single-line",
  "block-opening-brace-space-before": "always"
}
```

It's not possible to *enforce* differing styles for specific at-rules, but it is possible to *allow* differing styles using [these secondary options](https://github.com/stylelint/stylelint/blob/master/src/rules/block-closing-brace-newline-after/README.md#optional-options). See [this conversation](https://github.com/stylelint/stylelint/issues/561) about why we settled on this compromise. So, you can add the following options to the above config:

```
{
  "block-closing-brace-newline-after": ["always", { "ignoreAtRules: ["if", "else"] } ],
  "block-closing-brace-newline-before": "always-multi-line",
  "block-closing-brace-space-before": "always-single-line",
  "block-opening-brace-newline-after": "always-multi-line",
  "block-opening-brace-space-after": "always-single-line",
  "block-opening-brace-space-before": "always"
}
```

To allow:

```scss
@if (true) {
    // do something
} @else {
    // do something else
}
```

@ntwb It is spotting things correctly, it's just the [message is wrong](https://github.com/stylelint/stylelint/blob/master/src/rules/block-closing-brace-newline-after/index.js#L18). It should say "newline", rather than "single space" :) I'll patch it up now.

**Comment by Nirazul** [comment]
Ok, I was *totally* unaware of the possibility to duplicate rules.
This definitely should be written somewhere top, bold in all caps 👍

**Comment by jeddy3** [comment]
> Ok, I was totally unaware of the possibility to duplicate rules.

I'm not quite sure what you mean by "duplicate rules".  Can you elaborate please?

Each rule is only used once in the configuration. Each whitespace rule has a [very specific role](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md#whitespace-rules) and these rules can be used together in lots of different ways as to support a diverse set of style conventions.

**Comment by Nirazul** [comment]
Hah, my fault. I though that I've spotted the same rule twice 😬

**Comment by jeddy3** [comment]
It's OK :)

We've tried to find a balance between ease-of-use and flexibility (I was originally surprised at just how varied people format their CSS!). We gave it a lot of thought and found that standalone & narrowly-focused rules provided the best compromise. We've tried to document the system as much as possible (see the previous bits of documentation I've linked to e.g  [this](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md#whitespace-rules) and [this](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#rules-work-together)), but we're always looking for help in making the documentation better. If you have any ideas then please fire our a PR with those changes :)
